File: 29-DivideTwoIntegers/29-DivideTwoIntegers.py
# Last updated: 6/17/2025, 10:28:17 PM
class Solution:
    def divide(self, dividend: int, divisor: int) -> int:
        q = 0
        sign = 1
        if divisor == 1:
            return dividend
        if ((dividend <0) and (divisor >=0)) or ((dividend>=0) and (divisor<0)):
            sign = -1
        dd = abs(dividend)
        dr = abs(divisor)
        for i in range(30, -1, -1): 
            if dd >= (dr << i):
                q += (1 << i)
                dd -= (dr << i)
        return sign * q
            
        

File: 29-DivideTwoIntegers/29-DivideTwoIntegers.java
// Last updated: 9/4/2025, 10:32:22 AM
class Solution {
    public int divide(int dividend, int divisor) {
        if (dividend == Integer.MIN_VALUE && divisor == -1) {
            return Integer.MAX_VALUE;
        }
        int count = 0;
        boolean positive = (dividend >= 0 && divisor >= 0) || (dividend < 0 && divisor < 0);
        long a = Math.abs((long) dividend);
        long b = Math.abs((long) divisor);
        while (a >= b) {
            long temp = b;
            int multiple = 1;

            // Double divisor until it exceeds a
            while (a >= (temp << 1)) {
                temp <<= 1;
                multiple <<= 1;
            }

            // Subtract largest chunk
            a -= temp;
            count += multiple;
        }
       if(positive){
        return count;
        }
       else{
        return -(count);
       }
    }
}